### ISSUE TITLE
MemoryUsage of Arc forgets to account for ret-counts

### ISSUE THREAD

**Issue by matklad** [wasmerio/loupe#17]
https://github.com/wasmerio/loupe/blob/3108424c91c286c12fd4c54ab16d85d2177684a8/crates/loupe/src/memory_usage/sync.rs#L46-L52

This accounts for sizeof `T` stored inside the Arc, as well as sizeof of the Arc itself (fat or thin pointer). However, it misses the two counters that are stored in the `ArcInner`. In parcicular this currently returns 8: `crate::size_of_val(&Arc::new(()))`. I'd expect 24 instead -- one `usize` for the pointer +  `2 * usize` for strong and weak count.

**Comment by Hywan** [comment]
I'm surprised that `mem::size_of_val` misses the size of the `ArcInner`. I need to take a look at it.

**Comment by matklad** [comment]
Yeah, it's a bit tricky. The only difference between `size_of` and `size_of_val` is in handling DST. 

For `Arc`, the `size_of` is just a single pointer to the ArcInner heap allocated box. There isn't a function which can tell you the size of this box itself, you'll have to hard-code it.

Well, I guess there *is* a way to get this size without hard-coding, by asking the allocator about it. This is what servo's analogue dose: 

https://github.com/servo/servo/blob/94414dfb9aa8d29318e4825a36e4fdd188fbb892/components/malloc_size_of/lib.rs

But that obviously requires support from the allocator.

**Comment by matklad** [comment]
That's completly unrelated to the issue, but, JFYI, https://github.com/rust-analyzer/rust-analyzer/issues/9309
